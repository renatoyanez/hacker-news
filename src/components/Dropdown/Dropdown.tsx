import { ReactElement, useState } from 'react';
import { Typography } from 'components';
import { filterOptions } from 'helpers/filters';
import styles from './dropdown.module.css';

export interface Option {
	id: number;
	name: string;
	image: string;
	value: string;
}

export interface DropdownProps {
	options: Option[];
	onChange: (value: string) => void;
	placeholder?: string;
}

const Dropdown = ({
	options,
	placeholder,
	onChange,
}: DropdownProps): ReactElement => {
	const getFilterLocalStorage = localStorage.getItem('filter');

	const filterName = filterOptions.find(
		filter => filter.value === getFilterLocalStorage,
	)?.name;

	const [showOptionList, setShowOptionList] = useState(false);
	const [defaultSelectText, setDefaultSelectText] = useState(
		filterName || placeholder || '',
	);

	const handleListDisplay = () => {
		setShowOptionList(prevState => !prevState);
	};

	const handleOptionClick = (value: string, name: string) => {
		setDefaultSelectText(name);
		onChange(value);
		setShowOptionList(false);
	};

	return (
		<div>
			<div className={styles.container}>
				<div
					data-testid="dropdown-component"
					onClick={handleListDisplay}
					className={
						showOptionList
							? `${styles.selectedText} active`
							: styles.selectedText
					}
				>
					<Typography data-testid="selected-option" type="textStyle4">
						{defaultSelectText}
					</Typography>
				</div>
				{showOptionList && (
					<ul className={styles.selectOptions}>
						{options.map(option => {
							return (
								<li
									key={option.id}
									className={styles.option}
									onClick={() => {
										handleOptionClick(option.value, option.name);
									}}
								>
									<div className={styles.optionContainer}>
										<img
											src={option.image}
											alt={option.name}
											className={styles.image}
										/>
										<Typography type="textStyle4">
											{option.name}
										</Typography>
									</div>
								</li>
							);
						})}
					</ul>
				)}
			</div>
		</div>
	);
};

export default Dropdown;
