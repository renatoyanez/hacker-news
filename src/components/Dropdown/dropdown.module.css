.container {
	display: inline-block;
	min-width: 250px;
	position: relative;
}

.selectedText {
	background-color: #fff;
	padding: 0.313rem 0.75rem 0.313rem 0.75rem;
	border-radius: 4px;
	border: solid 1px #2e2e2e;
	background-color: #fff;
}

.selectedText.active::after {
	top: 8px;
	border-color: transparent transparent black transparent;
}

/* .selectedText::after {
	border: solid black;
	border-width: 0 3px 3px 0;
	display: inline-block;
	padding: 3px;
} */

.selectOptions {
	position: absolute;
	width: 100%;
	box-shadow: 0 2px 2px 0 #dad8d8;
}

.option {
	list-style-type: none;
	padding: 0.313rem 0.75rem 0.313rem 0.75rem;
	background: #fff;
	cursor: pointer;
}

.option:hover {
	background-color: #eaeaea;
}

ul {
	margin: 0;
	padding: 0;
}

.image {
	width: 1.5rem;
	height: 1.5rem;
	margin-right: 0.8rem;
}

.optionContainer {
	display: flex;
	align-items: center;
}
