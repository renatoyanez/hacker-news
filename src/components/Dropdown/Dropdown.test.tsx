import { create } from 'react-test-renderer';
import {
	render,
	fireEvent,
	cleanup,
	screen,
} from '@testing-library/react';
import Dropdown, { DropdownProps, Option } from './Dropdown';

afterEach(cleanup);

const options: Option[] = [
	{
		id: 1,
		image: 'string',
		name: 'First',
		value: 'First',
	},
	{
		id: 2,
		image: 'string',
		name: 'Second',
		value: 'Second',
	},
];

const props: DropdownProps = {
	options,
	onChange: jest.fn(),
	placeholder: 'Placeholder',
};

describe('Dropdown Component', () => {
	it('renders correctly', () => {
		const tree = create(<Dropdown {...props} />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('renders placeholder when value is empty', () => {
		const { getByText } = render(<Dropdown {...props} />);
		const placeholder = getByText(props.placeholder as string);
		expect(placeholder).toBeTruthy();
	});

	it('should call onChange when the first option is selected', async () => {
		const mockedOnChange = jest.fn();
		const container = render(
			<Dropdown
				options={options}
				onChange={mockedOnChange}
				placeholder={props.placeholder}
			/>,
		);

		const dropdownComponent = container.queryByTestId(
			'dropdown-component',
		) as Element;

		expect(dropdownComponent).toBeDefined();
		expect(dropdownComponent).not.toBeNull();
		expect(mockedOnChange).toHaveBeenCalledTimes(0);

		fireEvent.click(dropdownComponent);

		fireEvent.keyDown(dropdownComponent.firstChild as Element, {
			key: 'ArrowDown',
		});

		expect(screen.getAllByText('First')[0]).toBeInTheDocument();

		await fireEvent.click(container.getAllByText('First')[0]);

		const selectedText = container.getByText(
			props.options[0].name as string,
		);
		expect(selectedText).toBeTruthy();
		expect(mockedOnChange).toHaveBeenCalledTimes(1);
	});

	it('should call onChange when multiple options are selected', async () => {
		const mockedOnChange = jest.fn();
		const container = render(
			<Dropdown
				options={options}
				onChange={mockedOnChange}
				placeholder={props.placeholder}
			/>,
		);

		const dropdownComponent = container.queryByTestId(
			'dropdown-component',
		) as Element;

		expect(dropdownComponent).toBeDefined();
		expect(dropdownComponent).not.toBeNull();
		expect(mockedOnChange).toHaveBeenCalledTimes(0);

		fireEvent.click(dropdownComponent);
		expect(screen.getAllByText('Second')[0]).toBeInTheDocument();

		await fireEvent.click(container.getAllByText('Second')[0]);

		expect(screen.getAllByText('Second')[0]).toBeInTheDocument();

		fireEvent.keyDown(dropdownComponent.firstChild as Element, {
			key: 'ArrowDown',
		});

		fireEvent.click(dropdownComponent);
		await fireEvent.click(container.getAllByText('First')[0]);

		expect(screen.getAllByText('First')[0]).toBeInTheDocument();
		expect(mockedOnChange).toHaveBeenCalledTimes(2);
	});
});
