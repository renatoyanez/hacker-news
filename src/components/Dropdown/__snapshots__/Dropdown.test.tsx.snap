// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Component renders correctly 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="selectedText"
      data-testid="dropdown-component"
      onClick={[Function]}
    >
      <p
        className="textStyle4"
      >
        Placeholder
      </p>
    </div>
  </div>
</div>
`;
